akamai/cache-client.go

```diff
@@ -0,0 +1,255 @@
+package akamai
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/jmhodges/clock"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/golang.org/x/crypto/ocsp"
+
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+const (
+	purgePath       = "/ccu/v2/queues/default"
+	timestampFormat = "20060102T15:04:05-0700"
+)
+
+type purgeRequest struct {
+	Objects []string `json:"objects"`
+}
+
+type purgeResponse struct {
+	HTTPStatus       int    `json:"httpStatus"`
+	Detail           string `json:"detail"`
+	EstimatedSeconds int    `json:"estimatedSeconds"`
+	PurgeID          string `json:"purgeId"`
+}
+
+// CachePurgeClient talks to the Akamai CCU REST API. It is safe to make concurrent
+// purge requests.
+type CachePurgeClient struct {
+	client       *http.Client
+	apiEndpoint  string
+	apiHost      string
+	apiScheme    string
+	clientToken  string
+	clientSecret string
+	accessToken  string
+	issuer       *x509.Certificate
+	retries      int
+	retryBackoff time.Duration
+	log          *blog.AuditLogger
+	stats        statsd.Statter
+	clk          clock.Clock
+}
+
+type errFatal string
+
+func (e errFatal) Error() string { return string(e) }
+
+var (
+	// ErrAllRetriesFailed lets the caller of Purge to know if all the purge submission
+	// attempts failed
+	ErrAllRetriesFailed = errors.New("All attempts to submit purge request failed")
+)
+
+// NewCachePurgeClient constructs a new CachePurgeClient
+func NewCachePurgeClient(
+	endpoint,
+	clientToken,
+	clientSecret,
+	accessToken string,
+	retries int,
+	retryBackoff time.Duration,
+	issuer *x509.Certificate,
+	log *blog.AuditLogger,
+) (*CachePurgeClient, error) {
+	apiURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	return &CachePurgeClient{
+		client:       new(http.Client),
+		apiEndpoint:  endpoint,
+		apiHost:      apiURL.Host,
+		apiScheme:    strings.ToLower(apiURL.Scheme),
+		clientToken:  clientToken,
+		clientSecret: clientSecret,
+		accessToken:  accessToken,
+		retries:      retries,
+		retryBackoff: retryBackoff,
+		log:          log,
+		clk:          clock.Default(),
+		issuer:       issuer,
+	}, nil
+}
+
+func (cpc *CachePurgeClient) constructAuthHeader(request *http.Request, body []byte, apiPath string, nonce string) (string, error) {
```

Comment:
Or better yet add a link to docs in the code. General comment: I'd like to see a lot more comments in this code.